server/docker_resource.py

```diff
@@ -0,0 +1,433 @@
+
+        cliData = {}
+        cliData[DockerCache.type] = type
+        cliData[DockerCache.xml] = xml
+
+        imageKey = self._getHashKey(img_name)
+        self.data[imageKey][DockerCache.cli_dict][cli_name] = cliData
+
+    def raw(self):
+        return self.data
+
+    def deleteImage(self, name):
+        imageKey = self._getHashKey(name)
+        if imageKey in self.data:
+            del self.data[imageKey]
+            return True
+        else:
+            return False
+
+    def _getHashKey(self, imgName):
+        imageKey = hashlib.sha256(imgName.encode()).hexdigest()
+        return imageKey
+
+    def getCLIXML(self, imgName, cli):
+        imgKey = self._getHashKey(imgName)
+        if imgKey in self.data:
+            imageData = self.data[imgKey]
+
+        else:
+            print('no image named %s' % imgName)
+            return None
+
+        if cli in imageData[DockerCache.cli_dict]:
+
+            return imageData[DockerCache.cli_dict][cli][DockerCache.xml]
+        else:
+            print('no cli named %s' % imgName)
+            return None
+
+    def getCLIDict(self, imgName):
+        cliDict = {}
+        imgKey = self._getHashKey(imgName)
+        if imgKey in self.data:
+            imgData = self.data[imgKey]
+            for (key, val) in iteritems(imgData[DockerCache.cli_dict]):
+                cliDict[key] = {DockerCache.type: val[DockerCache.type]}
+            return cliDict
+        else:
+            return None
+
+    def validate(self):
+        """
+        Enforce structure of the data, does not verify that xml and cli
+        types field values are appropriate
+        """
+        if self.data is {}:
+            return True
+        for (imgHash, imgData) in iteritems(self.data):
+            if DockerCache.imageName not in imgData:
+                raise DockerImageDataError('There is no key: '
+                                           '%s in the dict'
+                                           % DockerCache.imageName)
+            if imgHash != self._getHashKey(imgData[DockerCache.imageName]):
+                raise DockerImageDataError('The hash id is not the sha256 '
+                                           'hash of the name: %' %
+                                           imgData[DockerCache.imageName])
+            if DockerCache.cli_dict not in imgData:
+                raise DockerImageDataError('There is no dictionary of cli ')
+            for (cliName, cliData) in iteritems(imgData[DockerCache.cli_dict]):
+                if DockerCache.type not in cliData:
+                    raise DockerImageDataError('The cli %s does not have a key'
+                                               ' for the type of cli' % cliName)
+                if DockerCache.xml not in cliData:
+                    raise DockerImageDataError('The cli %s does not have a key '
+                                               'for the xml of cli' % cliName)
```

Comment:
I think all of this validation can be improved by writing a JSON template (or xml template for xml) and validating against that. But this is good for now.